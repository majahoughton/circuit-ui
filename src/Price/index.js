import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import withStyles from '../../util/withStyles';
import styles from './index.scss';
import { formatNumberParts } from '../../util/numbers';
import { getCurrencyFormat } from '../../util/currency';

const Price = ({ currency, locale, amount, installments, hasDisclaimer }) => {
  const {
    decimalSep,
    thousandSep,
    prepend,
    currencyPrecision,
    addSpace,
    symbol
  } = getCurrencyFormat(currency, locale);

  const { integer, fractional } = formatNumberParts(amount, {
    precision: currencyPrecision,
    thousandSep
  });

  const showInstallments = installments > 1;
  const ccySymbolClassNames = classNames('price__currency', {
    'price__small-text': prepend && showInstallments
  });
  const ccySymbol = <span className={ccySymbolClassNames}>{symbol}</span>;
  const inst = showInstallments && (
    <span className="price__small-text">{`${installments}\u00D7`}</span>
  );

  const preClassNames = classNames('price__pre', {
    'price__pre--installments': showInstallments
  });
  const pre = (prepend || showInstallments) && (
    <div className={preClassNames}>
      {inst}
      {prepend && ccySymbol}
    </div>
  );
  const post = !prepend && ccySymbol;
  const frac = fractional !== '00' && (
    <span className="price__fractional">{`${decimalSep}${fractional}`}</span>
  );

  return (
    <div className="price">
      {pre}
      <div
        className={classNames('amount', {
          'amount--prepended-space': addSpace && prepend,
          'amount--postpended-space': addSpace && !prepend
        })}
      >
        <span>{integer}</span>
        {frac}
        {post}
      </div>
      {hasDisclaimer && <span>*</span>}
    </div>
  );
};

Price.propTypes = {
  currency: PropTypes.string.isRequired,
  locale: PropTypes.string.isRequired,
  amount: PropTypes.number.isRequired,
  installments: PropTypes.number,
  hasDisclaimer: PropTypes.bool
};

Price.defaultProps = {
  fractionalPrecision: 2,
  installments: 1,
  hasDisclaimer: false
};

export default withStyles(styles)(Price);
