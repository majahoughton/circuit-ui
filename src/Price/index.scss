@import './../shared';

.price {
  color: $color-brand;
  display: flex;
  font-size: $font-size-xxxl;
  line-height: $line-height-l;
}

.amount {
  align-items: flex-start;
  display: flex;
}

.amount--prepended-space {
  margin-left: $spacing-xs;
}

.amount--postpended-space {
  margin-right: $spacing-xs;
}

.price--disclaimer {
  margin-bottom: $spacing-s;
}

.price__fractional {
  font-size: $font-size-s;
  line-height: $line-height-s;
}

.price__pre {
  display: inline-block;

  &.price__pre--installments {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
  }
}

.price__decorator-text {
  color: darken($color-medium, 30%);
}

.price__small-text {
  font-size: $font-size-s;
  font-weight: $font-weight-medium;
  line-height: 14px;
}
