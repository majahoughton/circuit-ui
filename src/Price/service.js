import { includes } from 'lodash/fp';

export const getColorClass = color => {
  const safeColor = color.toLowerCase();
  const DEFAULT_COLOR = 'brand';
  const VALID_OPTIONS = ['highlight', 'warning', 'error', DEFAULT_COLOR];
  const colorModifier = includes(safeColor, VALID_OPTIONS)
    ? safeColor
    : DEFAULT_COLOR;
  return `price--${colorModifier}`;
};
