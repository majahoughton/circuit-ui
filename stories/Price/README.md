# Price

This is our pricing components for marketing pages and the shop.