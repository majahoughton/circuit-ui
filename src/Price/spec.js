import React from 'react';
import renderer from 'react-test-renderer';
import Price from '.';

describe('Price', () => {
  it('should render a regular price by default.', () => {
    const price = renderer.create(
      <Price currency="EUR" locale="de-DE" amount={9.99} />
    );
    expect(price).toMatchSnapshot();
  });
  it('should prepend the currency symbol when necessary.', () => {
    const prependedSymbolPrice = renderer.create(
      <Price currency="USD" locale="en-US" amount={9.99} />
    );
    expect(prependedSymbolPrice).toMatchSnapshot();
  });
  it('should render installments when they are > 1.', () => {
    const installmentsPrice = renderer.create(
      <Price currency="EUR" locale="de-DE" installments={9} amount={9.99} />
    );
    expect(installmentsPrice).toMatchSnapshot();
  });
  it('should render an asterisk when hasDisclaimer is truthy.', () => {
    const disclaimerPrice = renderer.create(
      <Price currency="EUR" locale="de-DE" amount={9.99} hasDisclaimer={true} />
    );
    expect(disclaimerPrice).toMatchSnapshot();
  });
  it("should not show the fractional part if it's zero.", () => {
    const integerPrice = renderer.create(
      <Price currency="EUR" locale="de-DE" amount={99} />
    );
    expect(integerPrice).toMatchSnapshot();
  });
});
