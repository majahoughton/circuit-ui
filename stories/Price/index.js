import React from 'react';
import { storiesOf } from '@storybook/react';
import { withNotes } from '@storybook/addon-notes';

import withTests from '../../util/withTests';
import { Price } from '../..';
import README from './README.md';

storiesOf('Prices', module)
  .addDecorator(withTests('Price/spec.js'))
  .add(
    'Price',
    withNotes(README)(() => (
      <div>
        {/* Adding some vertical spacing between the children */}
        <style
          scoped
          dangerouslySetInnerHTML={{
            __html: `
        div {
          margin-bottom: 12px;
        }
      `
          }}
        />
        <Price currency="EUR" locale="de-DE" amount={9.99} />
        <Price currency="EUR" locale="de-DE" amount={9} />
        <Price currency="EUR" locale="de-DE" installments={9} amount={9.99} />
        <Price currency="BRL" locale="pt-BR" installments={12} amount={9.99} />
        <Price currency="USD" locale="en-US" amount={9.99} />
        <Price currency="USD" locale="en-US" amount={49} />
        <Price currency="USD" locale="en-US" amount={9999999} />
        <Price currency="BRL" locale="pt-BR" amount={9.99} />
        <Price currency="EUR" locale="de-AT" amount={9.99} color="error" />
        <Price currency="EUR" locale="de-AT" amount={9.99} color="warning" />
        <Price currency="EUR" locale="de-AT" amount={9.99} color="highlight" />
      </div>
    ))
  );
