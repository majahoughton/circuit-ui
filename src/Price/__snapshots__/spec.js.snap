// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Price should not show the fractional part if it's zero. 1`] = `
<div
  className="price"
>
  <div
    className="amount amount--postpended-space"
  >
    <span>
      99
    </span>
    <span
      className="price__currency"
    >
      €
    </span>
  </div>
</div>
`;

exports[`Price should prepend the currency symbol when necessary. 1`] = `
<div
  className="price"
>
  <div
    className="price__pre"
  >
    <span
      className="price__currency"
    >
      $
    </span>
  </div>
  <div
    className="amount"
  >
    <span>
      9
    </span>
    <span
      className="price__fractional"
    >
      .99
    </span>
  </div>
</div>
`;

exports[`Price should render a regular price by default. 1`] = `
<div
  className="price"
>
  <div
    className="amount amount--postpended-space"
  >
    <span>
      9
    </span>
    <span
      className="price__fractional"
    >
      ,99
    </span>
    <span
      className="price__currency"
    >
      €
    </span>
  </div>
</div>
`;

exports[`Price should render an asterisk when hasDisclaimer is truthy. 1`] = `
<div
  className="price"
>
  <div
    className="amount amount--postpended-space"
  >
    <span>
      9
    </span>
    <span
      className="price__fractional"
    >
      ,99
    </span>
    <span
      className="price__currency"
    >
      €
    </span>
  </div>
  <span>
    *
  </span>
</div>
`;

exports[`Price should render installments when they are > 1. 1`] = `
<div
  className="price"
>
  <div
    className="price__pre price__pre--installments"
  >
    <span
      className="price__small-text"
    >
      9×
    </span>
  </div>
  <div
    className="amount amount--postpended-space"
  >
    <span>
      9
    </span>
    <span
      className="price__fractional"
    >
      ,99
    </span>
    <span
      className="price__currency"
    >
      €
    </span>
  </div>
</div>
`;
